Line detector documentation

iharwell opened the pull request:
Nothing too involved here. I went through and added XML documentation to nearly everything in the LineDetector project. This should make it a bit easier to work on moving forward.

There's a little bit left to do, though. In particular, some classes don't have class-level documentation yet.

There are also a couple of TODOs in the comments regarding some methods that should probably be replaced with properties, some interface members that could use some reorganizing, etc.

iharwell committed:
Three files documented.

iharwell committed:
ICrossingDetector documentation added.

iharwell committed:
Added documentation for the line set file format and the LineSets IO class.

iharwell committed:
Documented the ILineBasedDetector interface.

iharwell committed:
Documentation for FallingEdgeCrossingDetector added.

iharwell committed:
Updated documentation with linking.

iharwell committed:
Documentation for ISingleLineCrossingDetector.

iharwell committed:
Documentation for CascadedLinesDetector.

iharwell committed:
Documentation for the MultiLaneDetector.

iharwell committed:
Documentation for the SingleLineCrossingDetector class.

ghost commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/microsoft/Microsoft-Rocket-Video-Analytics-Platform?pullRequest=16) <br/>All CLA requirements met.

iharwell committed:
Fixed TODO formatting.

sharwell commented on src/VAP/LineDetector/CascadedLinesDetector.cs:
```diff
@@ -138,6 +153,12 @@ void NotifyCrossingEvent(int frameNo, int lineNo)
+        /// <param name="frameNo">The index of the frame to process.</param>
+        /// <param name="boxes">A list of bounding boxes of items in frame.</param>
+        /// <param name="mask">A mask detailing the precise layout of items in the frame using black to indicate vacant space, and white to indicate occupied space.</param>
         public void notifyFrameArrival(int frameNo, List<Box> boxes, Bitmap mask)
```
💡 Since this method implements `ILineBasedDetector.notifyFrameArrival`, the default behavior will operate as though the comment was:

```cs
/// <inheritdoc/>
```

Rather than duplicating documentation for implementations, it is often more efficient to document the original interface and allow that documentation to be inherited by the various implementations.

iharwell replied:
It is now updated with inheritdoc.

sharwell commented on src/VAP/LineDetector/CascadedLinesDetector.cs:
```diff
@@ -163,16 +189,28 @@ public void notifyFrameArrival(int frameNo, Bitmap mask)
+        /// <summary>
+        /// Gets the number of times that this detector has been triggered.
+        /// </summary>
+        /// <returns></returns>
```
💡 I would typically recommend omitting empty documentation elements. For this method, it's certainly easiest to omit `<returns>`; I would probably only add `<returns>` during an early documentation pass if there was a special value, e.g. if the method returned -1 in some cases to indicate "unknown count".

sharwell commented on src/VAP/LineDetector/CascadedLinesDetector.cs:
```diff
@@ -184,6 +222,10 @@ int getPendingNow(int frameNo, int lineNo)
         }
 
+        /// <summary>
+        /// Gets a <c>Dictionary</c> of the parameters used by this detector, stored by name.
```
💡 `see cref` is preferred for references to types.

```suggestion
        /// Gets a <see cref="Dictionary{TKey, TValue}"/> of the parameters used by this detector, stored by name.
```

sharwell commented on src/VAP/LineDetector/CascadedLinesDetector.cs:
```diff
@@ -193,11 +235,19 @@ int getPendingNow(int frameNo, int lineNo)
+        /// <summary>
+        /// Gets the current occupancy state of this detector. This updates when the detector is notified of a frame arrival.
+        /// </summary>
+        /// <returns>Returns true if the line is occupied, and false otherwise.</returns>
```
💡 The most common sentence structure I see for `<returns>` (or `<value>` for properties) is:

```suggestion
        /// <returns><see langword="true"/> if the line is occupied; otherwise, <see langword="false"/>.</returns>
```

Code completion handles `true` and `false`, so the fastest way to type this form is to type `true` followed by <kbd>Ctrl</kbd>+<kbd>Space</kbd>.

sharwell commented on src/VAP/LineDetector/DetectionLine.cs:
```diff
@@ -15,11 +15,33 @@ public class DetectionLine
+        /// <param name="a">The X coordinate of the first point of the line.</param>
+        /// <param name="b">The Y coordinate of the first point of the line.</param>
+        /// <param name="c">The X coordinate of the second point of the line.</param>
+        /// <param name="d">The Y coordinate of the second point of the line.</param>
```
💡 In cases where you have two overloads that are very similar, but one has an extra argument, you can use `<inheritdoc>` on the shorter signature to inherit the longer one:

```suggestion
        /// <inheritdoc cref="DetectionLine(int, int, int, int, double)"/>
```

This is certainly not required; if you expect the documentation to only be rarely updated then there is little downside to copying it. If you expect the documentation to be particularly long or vary over time, `<inheritdoc>` might provide an edge in maintainability.

iharwell replied:
It sounds like good practice, so I went ahead and changed it.

sharwell commented on src/VAP/LineDetector/Detector.cs:
```diff
@@ -34,6 +45,18 @@ public Detector(int sFactor, double rFactor, string linesFile, bool displayBGS)
+        /// <param name="fgmask">The foreground mask of the frame.</param>
+        /// <param name="boxes">A list of bounding boxes of items in the frame which deviate from the background.</param>
+        /// <returns>
+        ///   <para>Returns a <see cref="Tuple"/> with two <see cref="Dictionary{TKey, TValue}">Dictionaries</see>.</para>
```
💡 In this case, the tuple is backed by `ValueTuple<T1, T2>` as opposed to `Tuple<T1, T2>`. Probably easiest to write it without `see cref`.

```suggestion
        ///   <para>Returns a tuple with two <see cref="Dictionary{TKey, TValue}">Dictionaries</see>.</para>
```

sharwell commented on src/VAP/LineDetector/ICrossingDetector.cs:
```diff
@@ -1,9 +1,38 @@
+        /// <returns>Returns true if an event was detected, and false otherwise.</returns>
         bool notifyOccupancy(int frameNo, bool occupancy);
+
+        //TODO(isaach): Should be a property.
```
💡 If you do have a TODO with this form, consider matching the username form you use on GitHub

```suggestion
        //TODO(iharwell): Should be a property.
```

sharwell commented on src/VAP/LineDetector/DetectionLine.cs:
```diff
@@ -15,11 +15,33 @@ public class DetectionLine
+        /// <summary>
+        /// The first point of the line.
+        /// </summary>
+        public Point p1 { get; set; }
```
❔ While the fields were not marked `readonly`, it's not clear that was intentional. Is it possible to make these properties readonly?

```suggestion
        public Point p1 { get; }
```

sharwell commented on src/VAP/LineDetector/DetectionLine.cs:
```diff
@@ -15,11 +15,33 @@ public class DetectionLine
+        /// <summary>
+        /// The overlap threshold used to determine occupancy.
+        /// </summary>
+        public double overlapFractionThreshold { get; set; } = DEFAULT_OCCUPANCY_THRESHOLD;
```
💡 Consider removing this default value assignment, and updating the constructor to chain like this:

```cs
: this(a, b, c, d, DEFAULT_OCCUPANCY_THRESHOLD)
```

sharwell commented on src/VAP/LineDetector/DetectionLine.cs:
```diff
@@ -71,6 +109,15 @@ public double getFractionContainedInBox(Box b, Bitmap mask)
+        /// </summary>
+        /// <param name="mask">A mask detailing the precise layout of items in the frame using black to indicate vacant space, and white to indicate occupied space.</param>
+        /// <returns>
+        /// Returns the fraction, from 0 to 1, of this <see cref="DetectionLine"/> that
+        /// overlaps the given mask with a value of 0 indicating no overlap and a value of 1
+        /// indicating complete overlap.
```
💭 This might read better?

> Returns the fraction of this `DetectionLine` that overlaps the given mask, from 0 indicating no overlap to 1 indicating complete overlap.

iharwell replied:
That's certainly better. I'll include the change in my next commit.

sharwell commented on src/VAP/LineDetector/DetectionLine.cs:
```diff
@@ -103,6 +150,12 @@ public double getFractionInForeground(Bitmap mask)
+        /// </summary>
+        /// <param name="boxes">The list of <see cref="Box"/> objects to check, representing the bounding boxes of items in frame.</param>
+        /// <param name="mask">A mask detailing the precise layout of items in the frame using black to indicate vacant space, and white to indicate occupied space.</param>
+        /// <returns>Returns a <see cref="Tuple{double, Box}"/> containing both the maximum overlap fraction found, and the <see cref="Box"/> associated with that overlap.</returns>
```
> `<see cref="Tuple{double, Box}"/>`

This isn't a supported form in documentation comments today. A request to add support for this was filed in https://github.com/dotnet/csharplang/issues/401, but has not been implemented so far.

iharwell replied:
I'll probably change that to <c> then, since it's an external reference.

sharwell reviewed (commented):
Submitting partial review comments

iharwell committed:
Addressing comments on pull request.

Minor changes to the documentation comments to ensure everything works as expected.

Minor refactoring removed to place in a separate branch.
